File: scripts/ai-workflow/phases/report.py
"""Phase 7: レポート作成フェーズ

Phase 1-6の成果物を統合し、最終レポートを作成する。
エグゼクティブサマリー、詳細な変更内容、マージチェックリスト、リスク評価を含む。
"""
from pathlib import Path
from typing import Dict, Any, List
from .base_phase import BasePhase


class ReportPhase(BasePhase):
    """レポート作成フェーズ"""

    def __init__(self, *args, **kwargs):
        """初期化"""
        super().__init__(
            phase_name='report',
            *args,
            **kwargs
        )

    def execute(self) -> Dict[str, Any]:
        """
        レポート作成フェーズを実行

        Returns:
            Dict[str, Any]: 実行結果
                - success: bool
                - output: str - report.mdのパス
                - error: Optional[str]
        """
        try:
            # Issue情報を取得
            issue_number = int(self.metadata.data['issue_number'])

            # 各フェーズの成果物パスを取得
            phase_outputs = self._get_phase_outputs(issue_number)

            # 必須フェーズの成果物が存在するか確認
            required_phases = ['requirements', 'design', 'test_scenario', 'implementation', 'test_result', 'documentation']
            for phase in required_phases:
                if not phase_outputs[phase].exists():
                    return {
                        'success': False,
                        'output': None,
                        'error': f'{phase}の成果物が見つかりません: {phase_outputs[phase]}'
                    }

            # Planning Phase成果物のパス取得
            planning_path_str = self._get_planning_document_path(issue_number)

            # 実行プロンプトを読み込み
            execute_prompt_template = self.load_prompt('execute')

            # working_dirからの相対パスを使用
            rel_paths = {}
            for phase_name, phase_path in phase_outputs.items():
                rel_paths[phase_name] = phase_path.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            execute_prompt = execute_prompt_template.replace(
                '{planning_document_path}',
                planning_path_str
            ).replace(
                '{requirements_document_path}',
                f'@{rel_paths["requirements"]}'
            ).replace(
                '{design_document_path}',
                f'@{rel_paths["design"]}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_paths["test_scenario"]}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_paths["implementation"]}'
            ).replace(
                '{test_implementation_document_path}',
                f'@{rel_paths["test_implementation"]}'
            ).replace(
                '{test_result_document_path}',
                f'@{rel_paths["test_result"]}'
            ).replace(
                '{documentation_update_log_path}',
                f'@{rel_paths["documentation"]}'
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=execute_prompt,
                max_turns=30,
                log_prefix='execute'
            )

            # report.mdのパスを取得
            output_file = self.output_dir / 'report.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'report.mdが生成されませんでした: {output_file}'
                }

            # GitHub Issueに成果物を投稿
            try:
                output_content = output_file.read_text(encoding='utf-8')
                self.post_output(
                    output_content=output_content,
                    title="最終レポート"
                )
            except Exception as e:
                print(f"[WARNING] 成果物のGitHub投稿に失敗しました: {e}")

            # ステータス更新: BasePhase.run()で実行されるため不要
            # self.metadata.update_phase_status('report', 'completed', str(output_file))
            # self.post_progress('completed', f'レポートが完了しました: {output_file.name}')

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            # ステータス更新: 失敗
            self.metadata.update_phase_status('report', 'failed')
            # BasePhase.run()で実行されるため不要
            # self.post_progress('failed', f'レポートが失敗しました: {str(e)}')

            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def review(self) -> Dict[str, Any]:
        """
        レポートをレビュー

        Returns:
            Dict[str, Any]: レビュー結果
                - result: str - PASS/PASS_WITH_SUGGESTIONS/FAIL
                - feedback: str
                - suggestions: List[str]
        """
        try:
            # report.mdを読み込み
            report_file = self.output_dir / 'report.md'

            if not report_file.exists():
                return {
                    'result': 'FAIL',
                    'feedback': 'report.mdが存在しません。',
                    'suggestions': ['execute()を実行してreport.mdを生成してください。']
                }

            # 各フェーズの成果物パス
            issue_number = int(self.metadata.data['issue_number'])
            phase_outputs = self._get_phase_outputs(issue_number)

            # レビュープロンプトを読み込み
            review_prompt_template = self.load_prompt('review')

            # working_dirからの相対パスを使用
            rel_path_report = report_file.relative_to(self.claude.working_dir)
            rel_paths = {}
            for phase_name, phase_path in phase_outputs.items():
                rel_paths[phase_name] = phase_path.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            review_prompt = review_prompt_template.replace(
                '{report_document_path}',
                f'@{rel_path_report}'
            ).replace(
                '{requirements_document_path}',
                f'@{rel_paths["requirements"]}'
            ).replace(
                '{design_document_path}',
                f'@{rel_paths["design"]}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_paths["test_scenario"]}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_paths["implementation"]}'
            ).replace(
                '{test_implementation_document_path}',
                f'@{rel_paths["test_implementation"]}'
            ).replace(
                '{test_result_document_path}',
                f'@{rel_paths["test_result"]}'
            ).replace(
                '{documentation_update_log_path}',
                f'@{rel_paths["documentation"]}'
            )

            # Claude Agent SDKでレビューを実行
            messages = self.execute_with_claude(
                prompt=review_prompt,
                max_turns=30,
                log_prefix='review'
            )

            # レビュー結果をパース
            review_result = self._parse_review_result(messages)

            # レビュー結果をファイルに保存
            review_file = self.review_dir / 'result.md'
            review_file.write_text(review_result['feedback'], encoding='utf-8')
            print(f"[INFO] レビュー結果を保存: {review_file}")

            # GitHub Issueにレビュー結果を投稿: BasePhase.run()で実行されるため不要
            # self.post_review(
            #     result=review_result['result'],
            #     feedback=review_result['feedback'],
            #     suggestions=review_result.get('suggestions')
            # )

            return review_result

        except Exception as e:
            return {
                'result': 'FAIL',
                'feedback': f'レビュー中にエラーが発生しました: {str(e)}',
                'suggestions': []
            }

    def revise(self, review_feedback: str) -> Dict[str, Any]:
        """
        レビュー結果を元にレポートを修正

        Args:
            review_feedback: レビュー結果のフィードバック

        Returns:
            Dict[str, Any]: 修正結果
                - success: bool
                - output: str - report.mdのパス
                - error: Optional[str]
        """
        try:
            # 元のレポートを読み込み
            report_file = self.output_dir / 'report.md'

            if not report_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': 'report.mdが存在しません。'
                }

            # 各フェーズの成果物パス
            issue_number = int(self.metadata.data['issue_number'])
            phase_outputs = self._get_phase_outputs(issue_number)

            # 修正プロンプトを読み込み
            revise_prompt_template = self.load_prompt('revise')

            # working_dirからの相対パスを使用
            rel_path_report = report_file.relative_to(self.claude.working_dir)
            rel_paths = {}
            for phase_name, phase_path in phase_outputs.items():
                rel_paths[phase_name] = phase_path.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            revise_prompt = revise_prompt_template.replace(
                '{report_document_path}',
                f'@{rel_path_report}'
            ).replace(
                '{review_feedback}',
                review_feedback
            ).replace(
                '{requirements_document_path}',
                f'@{rel_paths["requirements"]}'
            ).replace(
                '{design_document_path}',
                f'@{rel_paths["design"]}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_paths["test_scenario"]}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_paths["implementation"]}'
            ).replace(
                '{test_implementation_document_path}',
                f'@{rel_paths["test_implementation"]}'
            ).replace(
                '{test_result_document_path}',
                f'@{rel_paths["test_result"]}'
            ).replace(
                '{documentation_update_log_path}',
                f'@{rel_paths["documentation"]}'
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=revise_prompt,
                max_turns=30,
                log_prefix='revise'
            )

            # report.mdのパスを取得
            output_file = self.output_dir / 'report.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': '修正されたreport.mdが生成されませんでした。'
                }

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def _get_phase_outputs(self, issue_number: int) -> Dict[str, Path]:
        """
        各フェーズの成果物パスを取得

        Args:
            issue_number: Issue番号

        Returns:
            Dict[str, Path]: フェーズ名 → 成果物パス
        """
        base_dir = self.metadata.workflow_dir.parent / f'issue-{issue_number}'

        return {
            'requirements': base_dir / '01_requirements' / 'output' / 'requirements.md',
            'design': base_dir / '02_design' / 'output' / 'design.md',
            'test_scenario': base_dir / '03_test_scenario' / 'output' / 'test-scenario.md',
            'implementation': base_dir / '04_implementation' / 'output' / 'implementation.md',
            'test_implementation': base_dir / '05_test_implementation' / 'output' / 'test-implementation.md',
            'test_result': base_dir / '06_testing' / 'output' / 'test-result.md',
            'documentation': base_dir / '07_documentation' / 'output' / 'documentation-update-log.md'
        }

File: scripts/ai-workflow-v2/src/phases/report.ts
import fs from 'fs-extra';
import path from 'node:path';
import { BasePhase, type PhaseInitializationParams, type PhaseRunOptions } from './base-phase.js';
import { PhaseExecutionResult } from '../types.js';

type PhaseOutputInfo = {
  path: string;
  exists: boolean;
};

type PhaseOutputMap = Record<string, PhaseOutputInfo>;

export class ReportPhase extends BasePhase {
  constructor(params: PhaseInitializationParams) {
    super({ ...params, phaseName: 'report' });
  }

  public async run(options: PhaseRunOptions = {}): Promise<boolean> {
    // 親クラスの run() を実行（execute + review cycle）
    const success = await super.run(options);

    // すべての処理が成功した場合のみ、ログをクリーンアップ（Issue #411）
    if (success) {
      const gitManager = options.gitManager ?? null;
      const issueNumber = parseInt(this.metadata.data.issue_number, 10);
      try {
        await this.cleanupWorkflowLogs(issueNumber);
        console.info('[INFO] Workflow logs cleaned up successfully.');

        // ログクリーンナップによる削除をコミット・プッシュ（Issue #411）
        if (gitManager) {
          await this.autoCommitAndPush(gitManager, null);
          console.info('[INFO] Cleanup changes committed and pushed.');
        }
      } catch (error) {
        const message = (error as Error).message ?? String(error);
        console.warn(`[WARNING] Failed to cleanup workflow logs: ${message}`);
      }
    }

    return success;
  }

  protected async execute(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const planningReference = this.getPlanningDocumentReference(issueNumber);

    // オプショナルコンテキストを構築（Issue #398）
    const requirementsContext = this.buildOptionalContext(
      'requirements',
      'requirements.md',
      '要件定義書は利用できません。Issue情報から要件を推測してください。',
      issueNumber,
    );

    const designContext = this.buildOptionalContext(
      'design',
      'design.md',
      '設計書は利用できません。Issue情報から設計内容を推測してください。',
      issueNumber,
    );

    const implementationContext = this.buildOptionalContext(
      'implementation',
      'implementation.md',
      '実装ログは利用できません。リポジトリの実装内容を確認してください。',
      issueNumber,
    );

    const testingContext = this.buildOptionalContext(
      'testing',
      'test-result.md',
      'テスト結果は利用できません。実装内容から推測してください。',
      issueNumber,
    );

    const documentationContext = this.buildOptionalContext(
      'documentation',
      'documentation-update-log.md',
      'ドキュメント更新ログは利用できません。',
      issueNumber,
    );

    // 参考情報（オプショナル）
    const scenarioContext = this.buildOptionalContext('test_scenario', 'test-scenario.md', '', issueNumber);
    const testImplementationContext = this.buildOptionalContext(
      'test_implementation',
      'test-implementation.md',
      '',
      issueNumber,
    );

    const executePrompt = this.loadPrompt('execute')
      .replace('{planning_document_path}', planningReference)
      .replace('{requirements_context}', requirementsContext)
      .replace('{design_context}', designContext)
      .replace('{implementation_context}', implementationContext)
      .replace('{testing_context}', testingContext)
      .replace('{documentation_context}', documentationContext)
      .replace('{test_scenario_context}', scenarioContext)
      .replace('{test_implementation_context}', testImplementationContext)
      .replace('{issue_number}', String(issueNumber));

    await this.executeWithAgent(executePrompt, { maxTurns: 30 });

    const reportFile = path.join(this.outputDir, 'report.md');
    if (!fs.existsSync(reportFile)) {
      return {
        success: false,
        error: `report.md が見つかりません: ${reportFile}`,
      };
    }

    // Phase outputはPRに含まれるため、Issue投稿は不要（Review resultのみ投稿）
    // try {
    //   const content = fs.readFileSync(reportFile, 'utf-8');
    //   await this.postOutput(content, '最終レポート');
    // } catch (error) {
    //   const message = (error as Error).message ?? String(error);
    //   console.warn(`[WARNING] GitHub へのレポート投稿に失敗しました: ${message}`);
    // }

    const outputs = this.getPhaseOutputs(issueNumber);
    await this.updatePullRequestSummary(issueNumber, outputs);

    return {
      success: true,
      output: reportFile,
    };
  }

  protected async review(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const reportFile = path.join(this.outputDir, 'report.md');

    if (!fs.existsSync(reportFile)) {
      return {
        success: false,
        error: 'report.md が存在しません。execute() を先に実行してください。',
      };
    }

    const reviewPrompt = this.buildPrompt('review', issueNumber, reportFile);

    const messages = await this.executeWithAgent(reviewPrompt, { maxTurns: 30, logDir: this.reviewDir });
    const reviewResult = await this.contentParser.parseReviewResult(messages);

    const reviewFile = path.join(this.reviewDir, 'result.md');
    fs.writeFileSync(reviewFile, reviewResult.feedback, 'utf-8');

    await this.github.postReviewResult(
      issueNumber,
      this.phaseName,
      reviewResult.result,
      reviewResult.feedback,
      reviewResult.suggestions,
    );

    return {
      success: reviewResult.result !== 'FAIL',
      output: reviewResult.result,
      error: reviewResult.result === 'FAIL' ? reviewResult.feedback : undefined,
    };
  }

  public async revise(reviewFeedback: string): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const reportFile = path.join(this.outputDir, 'report.md');

    if (!fs.existsSync(reportFile)) {
      return {
        success: false,
        error: 'report.md が存在しません。execute() を先に実行してください。',
      };
    }

    const revisePrompt = this.buildPrompt('revise', issueNumber, reportFile).replace(
      '{review_feedback}',
      reviewFeedback,
    );

    await this.executeWithAgent(revisePrompt, { maxTurns: 30, logDir: this.reviseDir });

    if (!fs.existsSync(reportFile)) {
      return {
        success: false,
        error: '改訂後の report.md を確認できませんでした。',
      };
    }

    return {
      success: true,
      output: reportFile,
    };
  }

  private buildPrompt(promptType: 'review' | 'revise', issueNumber: number, reportPath: string): string {
    const reportReference = this.getAgentFileReference(reportPath);
    if (!reportReference) {
      throw new Error(`Failed to compute reference for ${reportPath}`);
    }

    const planningReference = this.getPlanningDocumentReference(issueNumber);

    // オプショナルコンテキストを構築（Issue #398）
    const requirementsContext = this.buildOptionalContext('requirements', 'requirements.md', '', issueNumber);
    const designContext = this.buildOptionalContext('design', 'design.md', '', issueNumber);
    const implementationContext = this.buildOptionalContext('implementation', 'implementation.md', '', issueNumber);
    const testingContext = this.buildOptionalContext('testing', 'test-result.md', '', issueNumber);
    const documentationContext = this.buildOptionalContext(
      'documentation',
      'documentation-update-log.md',
      '',
      issueNumber,
    );
    const scenarioContext = this.buildOptionalContext('test_scenario', 'test-scenario.md', '', issueNumber);
    const testImplementationContext = this.buildOptionalContext(
      'test_implementation',
      'test-implementation.md',
      '',
      issueNumber,
    );

    return this.loadPrompt(promptType)
      .replace('{report_document_path}', reportReference)
      .replace('{planning_document_path}', planningReference)
      .replace('{requirements_context}', requirementsContext)
      .replace('{design_context}', designContext)
      .replace('{implementation_context}', implementationContext)
      .replace('{testing_context}', testingContext)
      .replace('{documentation_context}', documentationContext)
      .replace('{test_scenario_context}', scenarioContext)
      .replace('{test_implementation_context}', testImplementationContext)
      .replace('{issue_number}', String(issueNumber));
  }

  private getPhaseOutputs(issueNumber: number): PhaseOutputMap {
    const baseDir = path.resolve(this.metadata.workflowDir, '..', `issue-${issueNumber}`);

    const entries: Record<string, string> = {
      requirements: path.join(baseDir, '01_requirements', 'output', 'requirements.md'),
      design: path.join(baseDir, '02_design', 'output', 'design.md'),
      test_scenario: path.join(baseDir, '03_test_scenario', 'output', 'test-scenario.md'),
      implementation: path.join(baseDir, '04_implementation', 'output', 'implementation.md'),
      test_implementation: path.join(
        baseDir,
        '05_test_implementation',
        'output',
        'test-implementation.md',
      ),
      test_result: path.join(baseDir, '06_testing', 'output', 'test-result.md'),
      documentation: path.join(
        baseDir,
        '07_documentation',
        'output',
        'documentation-update-log.md',
      ),
    };

    return Object.fromEntries(
      Object.entries(entries).map(([phase, filePath]) => [
        phase,
        {
          path: filePath,
          exists: fs.existsSync(filePath),
        },
      ]),
    );
  }

  private async updatePullRequestSummary(issueNumber: number, outputs: PhaseOutputMap): Promise<void> {
    try {
      let prNumber = this.metadata.data.pr_number ?? null;
      const branchName =
        this.metadata.data.branch_name ?? `ai-workflow/issue-${this.metadata.data.issue_number}`;

      if (!prNumber) {
        const existing = await this.github.checkExistingPr(branchName);
        if (existing) {
          prNumber = existing.pr_number;
        } else {
          console.info('[INFO] No existing PR found. Skipping PR body update.');
          return;
        }
      }

      if (!prNumber) {
        return;
      }

      const extractedInfo = await this.github.extractPhaseOutputs(
        issueNumber,
        Object.fromEntries(
          Object.entries(outputs).map(([key, value]) => [key, value.path]),
        ),
      );

      const prBody = this.github.generatePrBodyDetailed(issueNumber, branchName, extractedInfo);
      const result = await this.github.updatePullRequest(prNumber, prBody);

      if (!result.success) {
        console.warn(`[WARNING] Failed to update PR body: ${result.error ?? 'unknown error'}`);
      } else {
        console.info(`[INFO] Updated PR #${prNumber} summary.`);
      }
    } catch (error) {
      const message = (error as Error).message ?? String(error);
      console.warn(`[WARNING] Failed to update PR summary: ${message}`);
    }
  }

  /**
   * ワークフローログをクリーンアップ（Issue #405）
   *
   * Report完了後に実行され、各フェーズのexecute/review/reviseディレクトリを削除します。
   * metadata.jsonとoutput/*.mdファイルは保持されます。
   *
   * @param issueNumber - Issue番号
   */
  private async cleanupWorkflowLogs(issueNumber: number): Promise<void> {
    const baseDir = path.resolve(this.metadata.workflowDir, '..', `issue-${issueNumber}`);

    // Planning フェーズ（00_planning）は削除対象外（Issue参照ソースとして保持）
    const phaseDirectories = [
      '01_requirements',
      '02_design',
      '03_test_scenario',
      '04_implementation',
      '05_test_implementation',
      '06_testing',
      '07_documentation',
      '08_report',
    ];

    const targetSubdirs = ['execute', 'review', 'revise'];

    let deletedCount = 0;
    let skippedCount = 0;

    for (const phaseDir of phaseDirectories) {
      const phasePath = path.join(baseDir, phaseDir);

      if (!fs.existsSync(phasePath)) {
        skippedCount++;
        continue;
      }

      for (const subdir of targetSubdirs) {
        const subdirPath = path.join(phasePath, subdir);

        if (fs.existsSync(subdirPath)) {
          try {
            fs.removeSync(subdirPath);
            deletedCount++;
            console.info(`[INFO] Deleted: ${path.relative(baseDir, subdirPath)}`);
          } catch (error) {
            const message = (error as Error).message ?? String(error);
            console.warn(`[WARNING] Failed to delete ${subdirPath}: ${message}`);
          }
        }
      }
    }

    console.info(
      `[INFO] Cleanup summary: ${deletedCount} directories deleted, ${skippedCount} phase directories skipped.`,
    );
  }
}
